src/lib/Microsoft.Health.Fhir.Ingest/Data/DeterministicHashCodeGenerator.cs

```diff
@@ -0,0 +1,47 @@
+﻿// -------------------------------------------------------------------------------------------------
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License (MIT). See LICENSE in the repo root for license information.
+// -------------------------------------------------------------------------------------------------
+
+using System;
+using System.Security.Cryptography;
+using System.Text;
+using EnsureThat;
+using Microsoft.Extensions.Caching.Memory;
+
+namespace Microsoft.Health.Fhir.Ingest.Data
+{
+    public class DeterministicHashCodeGenerator : IHashCodeGenerator
+    {
+        private readonly HashAlgorithm _hasher;
+        private readonly IMemoryCache _hashCodeCache;
+        private readonly TimeSpan _cachedItemRetention = TimeSpan.FromMinutes(5);
+
+        public DeterministicHashCodeGenerator(IMemoryCache hashCodeCache)
+        {
+            _hashCodeCache = EnsureArg.IsNotNull(hashCodeCache, nameof(hashCodeCache));
+            _hasher = SHA256.Create();
+        }
+
+        public void Dispose()
+        {
+            _hasher.Dispose();
+            _hashCodeCache.Dispose();
+        }
+
+        public int GenerateHashCode(string valueToHash)
+        {
+            EnsureArg.IsNotNullOrEmpty(valueToHash, nameof(valueToHash));
+
+            return _hashCodeCache.GetOrCreate(
+                valueToHash,
+                e =>
+                {
+                    e.SetSlidingExpiration(_cachedItemRetention);
+                    e.Size = 1;
+                    var hash = _hasher.ComputeHash(Encoding.Unicode.GetBytes(valueToHash));
+                    return BitConverter.ToInt32(hash, 0);
```

Comment:
Sounds good thanks Rob! I was thinking of this and I wonder if we can do something more targeted for our use case.  We could go very simple and just look at the first character of the value (device id). The main goal of these code is have a fast and simple bucketing of a device to a  partition in a detrerministic way.  I am trying to think of a simple scheme that would allow almost a round robin distribution if device ids are assigned in a sequential manner.  Potentially over optimzing and fine keeping this but still wanted to discuss.

Reply:
I'm thinking its safer to hash over the entirety of the device id. What you're proposing could work if we could guarantee entropy in the first character of the device id. For example, if the customer used simple Guids as thier device id. But I'm concerned that that could lead to all measurements landing on a single partition if they don't use Guids. For example, if the customer creates device ids in the format `deviceIdXXXXX` where XXXXX is some incrementing number. If we simply take the first character it will always be `d`, which will force all measurements onto the same partition.